consensus/istanbul/backend/handler.go

```diff
@@ -290,3 +298,171 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
+	}()
+	go func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		errCh <- err
+		isValidatorCh <- isValidator
+	}()
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	// We write to errCh four times unless if the timeout is triggered or an error isn't nil
+	for i := 0; i < 4; i++ {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				return false, err
+			}
+		case <-timeout.C:
+			return false, p2p.DiscReadTimeout
+		}
+	}
+	return <-isValidatorCh, nil
+}
+
+// generateValidatorProofMessage will create a directAnnounce message that serves
+// as a proof to the peer during the handshake that this node is a validator.
+// If this node is not a validator or this node does not believe the peer
+// is a validator, no proof is generated to hide the address of this node.
+// If this node is a proxy, it will use the most recent directAnnounce this
+// node has received from the proxied validator.
+func (sb *Backend) generateValidatorProofMessage(peer consensus.Peer) (*istanbul.Message, error) {
+	shouldSend, err := sb.shouldSendValidatorProof(peer)
+	if err != nil {
+		return nil, err
+	}
+	if shouldSend {
+		if sb.config.Proxy {
+			// if this proxy has been disconnected from its validator for any reason,
+			// don't send a proof message, but do not fail the handshake.
+			if sb.proxyDirectAnnounceMsg != nil && sb.proxiedPeer != nil {
+				// Make a copy of the proxyDirectAnnounceMsg for thread safety
+				sb.proxyDirectAnnounceMsgMu.RLock()
+				defer sb.proxyDirectAnnounceMsgMu.RUnlock()
+				return sb.proxyDirectAnnounceMsg.Copy(), nil
+			}
+		} else {
+			return sb.generateDirectAnnounce(getCurrentAnnounceVersion())
+		}
+	}
+	// Even if we decide not to identify ourselves,
+	// send an empty message to complete the handshake
+	return &istanbul.Message{}, nil
+}
+
+// shouldSendValidatorProof determines if this node should send a
+// validator proof revealing its address to a peer
+func (sb *Backend) shouldSendValidatorProof(peer consensus.Peer) (bool, error) {
+	var validatorAddress common.Address
+	if sb.config.Proxy {
+		validatorAddress = sb.config.ProxiedValidatorAddress
+	} else {
+		validatorAddress = sb.Address()
+	}
+	// Check to see if this node is a validator
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		return false, err
+	}
+	if !regAndActiveVals[validatorAddress] {
+		return false, nil
+	}
+	// If the peer is not a known validator, we do not want to give up extra
+	// information about ourselves
+	if _, err := sb.valEnodeTable.GetAddressFromNodeID(peer.Node().ID()); err != nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// readValidatorProofMessage reads a validator proof message as a part of the
+// istanbul handshake. Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorProofMessage(peer consensus.Peer) (bool, error) {
+	peerMsg, err := peer.ReadMsg()
```

Comment:
Is the peer guaranteed to send a validator proof message here? Could we accidentally read a different message?